File: day11/main.rb
t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)

g = {}
File.read('input.txt').split("\n").each do |line|
  next if line.strip.empty?
  from, rest = line.split(':')
  g[from.strip] = rest.strip.split(/\s+/)
end

def count_paths(g, start, target)
  memo = Hash.new do |h, node|
    h[node] = if node == target
                1
              else
                (g[node] || []).sum { |nxt| h[nxt] }
              end
  end
  memo[start]
end

p1 = count_paths(g, 'you', 'out')
a1 = count_paths(g, 'svr', 'dac')
a2 = count_paths(g, 'dac', 'fft')
a3 = count_paths(g, 'fft', 'out')
b1 = count_paths(g, 'svr', 'fft')
b2 = count_paths(g, 'fft', 'dac')
b3 = count_paths(g, 'dac', 'out')
p2 = a1 * a2 * a3 + b1 * b2 * b3

elapsed_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000
puts "paths_you_to_out=#{p1} paths_svr_via_dac_fft=#{p2} elapsed_ms=#{'%.3f' % elapsed_ms}"

File: day11/Main.hs
{-# LANGUAGE BangPatterns #-}
module Main where

import qualified Data.Map.Strict as M
import Data.Char (isSpace)
import Data.IORef
import GHC.Clock (getMonotonicTimeNSec)
import Text.Printf (printf)

trim :: String -> String
trim = f . f where f = reverse . dropWhile isSpace

loadGraph :: [String] -> M.Map String [String]
loadGraph = foldl addLine M.empty
  where
    addLine m line =
      let t = trim line
      in if null t then m
         else let (l,r) = break (==':') t
                  src = trim l
                  dests = words (trim (drop 1 r))
              in M.insert src dests m

-- Use IORef for mutable memoization
countPaths :: M.Map String [String] -> String -> String -> IO Integer
countPaths g start target = do
    memoRef <- newIORef M.empty
    doneRef <- newIORef M.empty
    go memoRef doneRef start
  where
    go memoRef doneRef node = do
        done <- readIORef doneRef
        case M.lookup node done of
            Just _ -> do
                memo <- readIORef memoRef
                return $ M.findWithDefault 0 node memo
            Nothing -> do
                modifyIORef' doneRef (M.insert node True)
                if node == target
                    then do
                        modifyIORef' memoRef (M.insert node 1)
                        return 1
                    else do
                        let neighbors = M.findWithDefault [] node g
                        counts <- mapM (go memoRef doneRef) neighbors
                        let !total = sum counts
                        modifyIORef' memoRef (M.insert node total)
                        return total

main :: IO ()
main = do
    startTime <- getMonotonicTimeNSec
    ls <- lines <$> readFile "input.txt"
    let !g = loadGraph ls
    !p1 <- countPaths g "you" "out"
    !a1 <- countPaths g "svr" "dac"
    !a2 <- countPaths g "dac" "fft"
    !a3 <- countPaths g "fft" "out"
    !b1 <- countPaths g "svr" "fft"
    !b2 <- countPaths g "fft" "dac"
    !b3 <- countPaths g "dac" "out"
    let !p2 = a1*a2*a3 + b1*b2*b3
    endTime <- getMonotonicTimeNSec
    let elapsedMs = fromIntegral (endTime - startTime) / 1e6 :: Double
    printf "paths_you_to_out=%d paths_svr_via_dac_fft=%d elapsed_ms=%.3f\n" p1 p2 elapsedMs
